# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
#рассчитайем общее количество клиентов для каждого количества детей и общее количество клиентов,
#имевших задолженность по кредиту, для каждого количества детей.
debt_summary_by_children = pd.pivot_table(data, values='debt', index='children', aggfunc=['sum', 'count']) 
debt_summary_by_children.columns = ['Clients with Debt', 'Total Clients']
#для наглядности результаты представим в виде сводной таблицы, данные выведем в виде процентов.
debt_summary_by_children['Debt Percentage'] = ((debt_summary_by_children['Clients with Debt'] / debt_summary_by_children['Total Clients']) * 100).map('{:.2f}%'.format)
debt_summary_by_children

Clients with Debt  Total Clients Debt Percentage
children                                                  
0                      1063          14091           7.54%
1                       444           4808           9.23%
2                       194           2052           9.45%
3                        27            330           8.18%
4                         4             41           9.76%
5                         0              9           0.00%

**Вывод:** 
1. Категория без детей имеет наименьший процент клиентов с задолженностью - 7.54%, а категория с 4 детьми наибольший - 9.76%.
2. Процент клиентов с задолженностью увеличивается с увеличением числа детей с 1 до 2 (от 9.23% до 9.45% соответственно).
3. После этого, процент клиентов с задолженностью начинает снижаться с увеличением числа детей (8.18% для 3 детей). 
4. Однако стоит заметить, что количество данных для категорий с 4 и 5 детьми небольшое (41 и 9 клиентов соответственно), поэтому результаты для этих категорий могут быть менее репрезентативными.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [23]:
#Проверим, есть ли неявные дупликаты в данных
data['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

In [24]:
#рассчитайем общее количество клиентов для каждого типа семейного положения и общее количество клиентов,
#имевших задолженность по кредиту, для каждого типа семейного положения.
debt_summary_by_family_status = pd.pivot_table(data, values='debt', index='family_status', aggfunc=['sum', 'count'])
debt_summary_by_family_status.columns = ['Clients with Debt', 'Total Clients']
#для наглядности результаты представим в виде сводной таблицы, данные выведем в виде процентов.
debt_summary_by_family_status['Debt Percentage'] = ((debt_summary_by_family_status['Clients with Debt'] / debt_summary_by_family_status['Total Clients']) * 100).map('{:.2f}%'.format)
debt_summary_by_family_status.sort_values(by='Debt Percentage', ascending=False)

Clients with Debt  Total Clients Debt Percentage
family_status                                                          
Не женат / не замужем                273           2796           9.76%
гражданский брак                     385           4134           9.31%
женат / замужем                      927          12261           7.56%
в разводе                             84           1189           7.06%
вдовец / вдова                        63            951           6.62%

**Вывод:** 
1. Наибольший процент клиентов с задолженностью (9.76%) наблюдается среди клиентов, которые не женаты или не замужем.
2. Клиенты в гражданском браке также имеют высокий процент задолженности (9.31%).
3. Семейное положение "женат / замужем" имеет более низкий процент задолженности (7.56%).
4. Клиенты, находящиеся в разводе или являющиеся вдовцами / вдовами, имеют наименьший процент задолженности (7.06% и 6.62% соответственно).


#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [25]:
#рассчитайем общее количество клиентов для каждой категории дохода и общее количество клиентов,
#имевших задолженность по кредиту, для каждой категории дохода.
debt_summary_by_income = pd.pivot_table(data, values='debt', index='total_income_category', aggfunc=['sum', 'count'])
debt_summary_by_income.columns = ['Clients with Debt', 'Total Clients']
#для наглядности результаты представим в виде сводной таблицы, данные выведем в виде процентов.
debt_summary_by_income['Debt Percentage'] = ((debt_summary_by_income['Clients with Debt'] / debt_summary_by_income['Total Clients']) * 100).map('{:.2f}%'.format)
debt_summary_by_income.sort_values(by='total_income_category', ascending=False)

Clients with Debt  Total Clients Debt Percentage
total_income_category                                                  
E                                      2             22           9.09%
D                                     21            349           6.02%
C                                   1353          15921           8.50%
B                                    354           5014           7.06%
A                                      2             25           8.00%

**Вывод:** 
1. Категория E имеет самый высокий процент клиентов с задолженностью (9.09%).
2. Категории D и A имеют относительно небольшой процент клиентов с задолженностью (6.02% и 8.00% соответственно), что может указывать на более надежную кредитную платежеспособность в этих категориях дохода.
3. Категории B и C имеют промежуточные значения процента клиентов с задолженностью (7.06% и 8.50% соответственно).
4. Стоит отметить, что категории E, D и A имеют намного меньшее количество клиентов в сравнении с другими категориями, что может повлиять на статистическую значимость результатов. Выводы на основании этих групп не являются надежными.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [26]:
#рассчитайем общее количество клиентов для каждой цели кредита и общее количество клиентов,
#имевших задолженность по кредиту, для цели кредита.
debt_summary_by_purpose = pd.pivot_table(data, values='debt', index='purpose_category', aggfunc=['sum', 'count'])
debt_summary_by_purpose.columns = ['Clients with Debt', 'Total Clients']
#для наглядности результаты представим в виде сводной таблицы, данные выведем в виде процентов.
debt_summary_by_purpose['Debt Percentage'] = ((debt_summary_by_purpose['Clients with Debt'] / debt_summary_by_purpose['Total Clients']) * 100).map('{:.2f}%'.format)
debt_summary_by_purpose.sort_values(by='Debt Percentage', ascending=False)

Clients with Debt  Total Clients Debt Percentage
purpose_category                                                          
операции с автомобилем                  400           4279           9.35%
получение образования                   369           3988           9.25%
проведение свадьбы                      183           2313           7.91%
операции с недвижимостью                780          10751           7.26%

**Вывод:** 
1. Категории "операции с автомобилем" и "получение образования" имеют самый высокий процент клиентов с задолженностью (9.35% и 9.25% соответственно).
2. Категории "проведение свадьбы" и "операции с недвижимостью" имеют более низкий процент клиентов с задолженностью (7.91% и 7.26% соответственно).
3. Категория "операции с недвижимостью" имеет самый низкий процент клиентов с задолженностями при значительно более высоком количестве клиентов по сравнению с другими категориями, что делает данную категорию наиболее привлекательной.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 
Появление пропусков в столбцах "days_employed" и "total_income" может быть обусловлено несколькими причинами:
1. В некоторых случаях, клиенты могут не предоставлять информацию о своем трудовом стаже (days_employed) или общем доходе (total_income) при заполнении заявки на кредит. Это может быть связано с желанием сохранить свои персональные данные или просто потому, что эти поля не обязательны для заполнения.
2. Пропуски могут возникнуть из-за технических ошибок или сбоев в процессе заполнения данных. 
3. В некоторых случаях, пропуски могут быть связаны с ошибками в обработке данных сотрудниками кредитной организации при внесении информации в базу данных.

# 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* <br>
1. Медиана является устойчивой статистикой, которая не зависит от экстремальных значений (выбросов) в данных. Если в столбце есть выбросы, то среднее значение (среднее арифметическое) может существенно исказиться, в то время как медиана останется более стабильной и отражает центральную тенденцию данных.
2. Вычисление медианы - это относительно простая и быстрая операция, особенно когда имеется большой объем данных. 

### Шаг 4: общий вывод.

#### Предобработка данных
* В столбцах <b>days_employed</b>(общий трудовой стаж в днях) и <b>total_income</b>(ежемесячный доход) присутствовали пропуски. Недостающие значения заполнили медианным значением по типу занятости.
* В данных встречались отрицательные значения количества дней трудового стажа в столбце <b>days_employed</b>. Заменили все отрицательные значения положительными, причем для двух типов (безработные и пенсионеры) получатся аномально большие значения.
* В столбце <b>children</b> присутстовали два аномальных значения. Удалили строки, в которых встречаются такие аномальные значения из датафрейма.
* Заменили вещественный тип данных в столбце <b>total_income</b>(ежемесячный доход) на целочисленный.
* Удалили неявные дубликаты в столбце education(уровень образования клиента), приведя все к нижнему регистру. 
* Для категоризации данных создали в датафрейме столбец <b>total_income_category</b> с категориями:
    * 0–30000 — 'E';
    * 30001–50000 — 'D';
    * 50001–200000 — 'C';
    * 200001–1000000 — 'B';
    * 1000001 и выше — 'A'.

#### По предоставленной таблице с данными о клиентах и задолженностях, можно сделать следующие общие выводы:
1. *По количеству детей*:
    * Категория без детей имеет наименьший процент клиентов с задолженностью - <b>7.54%</b>, а категория с **4 детьми** наибольший - <b>9.76%</b>.Это может указывать на то, что наличие детей может повышать финансовую нагрузку и повлиять на способность клиентов выплачивать кредиты в срок.
2. *По семейному статусу*:
    * Клиенты состоящие(процент клиентов с просрочкой - <b>7.56%</b>) или состоявшие в браке(в разводе - <b>7.06%</b>, вдовец / вдова - <b>6.62%</b>) реже допускают просрочку. Категория "не женат / не замужем" напротив имеет наибольший процент клиентов с просрочкой - <b>9.76%</b>. Это может указывать на то, что семейный статус может оказывать положительное влияние на платежеспособность и финансовую дисциплину клиентов.
3. *По целям кредитования*:
    * Категории "операции с автомобилем" и "получение образования" имеют наибольший процент клиентов с задолженностью (<b>9.35%</b> и <b>9.25%</b> соответственно). Категория "операции с недвижимостью" имеет самый низкий процент задолженности (<b>7.26%</b>)
4. *По категориям дохода*:
    * Категория **"B"** имеет относительно низкий процент задолженности (<b>7.06%</b>), в то время как категория **"C"** имеет большой уровень задолженности (<b>8.50%</b>). Данных по категорям **"A"**, **"D"**, **"E"** недостаточно для построяния каких-либо выводов.



<div style="font-size: 15px; padding-top: 50px">Таким образом, на основе предоставленных данных можно предположить, что семейное положение и количество детей клиента могут влиять на факт погашения кредита в срок. Клиенты, не состоящие в официальном браке и имеющие больше детей, имеют более высокий процент задолженности. 
</div>